docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,253 @@
+dependency based system that allows the authors of services to control
+the initialization and shutdown order of components using an
+dependencies to control the ordering.
+
+Dependency based orderings make it possible to write components
+(reactors, etc.) with only limited awareness of other components or of
+node initialization, and is the state of the art for process
+initialization. However, this may be too abstract and complicated, and
+authors of components in the current implementation of tendermint
+*would* need to know about other components, so a dependency based
+system would be unhelpfully abstract at this stage.
+
+## Decisions
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects.
+
+  - Reactors should not need to expose their interfaces *within* the
+	implementation of the node type, except in the context of some groups of
+	services.
+
+  - This refactoring should be entirely opaque to users.
+
+  - These node initialization changes should not require a
+	reevaluation of the `service.Service` or a generic initialization
+	orchestration framework.
+
+- If required for the 0.35 release, add an "unsafe" way to construct a
+  node service with user-supplied mempool, including making the
+  relevant interfaces and types external/exported.
+
+- Prioritize implementation of p2p-based statesync reactor to obviate
+  need for users to inject a custom state-sync provider.
+
+## Detailed Design
+
+The [current
+nodeImpl](https://github.com/tendermint/tendermint/blob/master/node/node.go#L47)
+includes direct references to the implementations of each of the
+reactors, which should be replaced by references to `service.Service`
+objects. This will require moving the construction of the [rpc
+service](https://github.com/tendermint/tendermint/blob/master/node/node.go#L771)
+into the constructor of
+[makeNode](https://github.com/tendermint/tendermint/blob/master/node/node.go#L126).
+
+In order to prevent adding complexity to the `node` package, this
+project will add an implementation to the service `service` package
+that implements `service.Service` and is comprised of a sequence of
+underlying `service.Service` objects and handles their
+startup/shutdown in a consistent order.
+
+Consensus, blocksync (nee fast sync.), and statesync all depend on
+each other, and have significant initialization dependencies that are
+presently encoded in the `node` package, and as part of this change,
+a new package/component will encapsulate the initialization of this
+area of functionality.
+
+In order to support replacement of a component, a new public function
+will be added to the public interface of `node` with a signature that
+resembles the following:
+
+```go
+func NewWithServices(conf *config.Config,
+	logger log.Logger,
+	cf proxy.ClientCreator,
+	gen *types.GenesisDoc,
+	srvs ...service.Service
```

Comment:
I advise caution about using variadic function signatures in a public API. If it's practical to do so, I'd avoid it.

It's not a correctness issue, but I have found it's harder for future readers to audit the correctness of variadic function call sites, especially when the API changes over time. It also complicates static analysis—or even just grepping for usage. We don't do much static analysis right now, but I would like to be able to: Correctness is really crucial in the consensus plumbing.

I see that we already have several variadic type signatures in this codebase, but I've been bitten by this enough times that I wanted to at least mention it. I would just make this an explicit slice, even if it does mean a little bit of extra typing at the call site.

Reply:
It's definitely practical. I don't have a strong feeling on the subject, I think as long as the type isn't `interface{}`... but I'm happy to change it.